misc/logger/src/circularArray.ts

```diff
@@ -0,0 +1,71 @@
+export default class CircularArray {
+  private array: string[];
+  private head: number;
+  private currentSizeInBytes: number;
+  private capacityInBytes: number;
+
+  constructor(capacityInBytes: number) {
+    this.capacityInBytes = capacityInBytes;
+    this.currentSizeInBytes = 0;
+    this.head = 0;
+    this.array = [];
+  }
+
+  private calculateStringSize(item: string): number {
+    return new TextEncoder().encode(item).length;
+  }
+
+  public enqueue(item: string): void {
+    const itemSize = this.calculateStringSize(item);
+    if (itemSize > this.capacityInBytes) {
+      throw new Error("Item size exceeds total capacity.");
+    }
+
+    let insertIdx = this.array.length;
+    while (this.currentSizeInBytes + itemSize > this.capacityInBytes) {
+      const headItem = this.array[this.head];
+      this.currentSizeInBytes -= this.calculateStringSize(headItem);
+      insertIdx = this.head;
```

Comment:
Seems like there is disagreement on if insertions should go into the middle of the array or the end. If the capacity gets too big, we will start overwriting the middle of the array. But then if it gets small enough it will start adding to the end. But these will then be out-of-order?

Reply:
Hmm maybe linked list would've been better for this situation 🤔